programs/mango-v4/src/accounts_ix/openbook_v2_liq_force_cancel_orders.rs

```diff
@@ -0,0 +1,81 @@
+use anchor_lang::prelude::*;
+use anchor_spl::token::{Token, TokenAccount};
+
+use crate::error::*;
+use crate::state::*;
+use openbook_v2::{
+    program::OpenbookV2,
+    state::{Market, OpenOrdersAccount},
+};
+
+#[derive(Accounts)]
+pub struct OpenbookV2LiqForceCancelOrders<'info> {
+    #[account(
+        constraint = group.load()?.is_ix_enabled(IxGate::OpenbookV2LiqForceCancelOrders) @ MangoError::IxIsDisabled,
+    )]
+    pub group: AccountLoader<'info, Group>,
+
+    // Allow force cancel even if account is frozen
+    #[account(
+        mut,
+        has_one = group
+    )]
+    pub account: AccountLoader<'info, MangoAccountFixed>,
+
+    #[account(
+        mut,
+        constraint = open_orders.load()?.market == openbook_v2_market_external.key(),
+    )]
+    pub open_orders: AccountLoader<'info, OpenOrdersAccount>,
+
+    #[account(
+        has_one = group,
+        has_one = openbook_v2_program,
+        has_one = openbook_v2_market_external,
+    )]
+    pub openbook_v2_market: AccountLoader<'info, OpenbookV2Market>,
+
+    pub openbook_v2_program: Program<'info, OpenbookV2>,
+
+    #[account(
+        has_one = bids,
+        has_one = asks,
+        has_one = event_queue,
+        constraint = openbook_v2_market_external.load()?.base_vault == market_base_vault.key(),
+        constraint = openbook_v2_market_external.load()?.quote_vault == market_quote_vault.key(),
```

Comment:
these are typically better as `address = openbook_v2_market_external.load()?.base_vault` style constraints on the target accounts. However, I'm not yet sure it's good to propagate all these constraints to the mango interface. We don't do that for serum.

The main pros are, imo, that it helps the anchor clients to autodeduce more accounts, and we might give better error messages to users. However, as a drawback, a bunch of openbook v2 specifics leak into mango and all these checks are done twice (once in the mango program and once when cpi-ing into obv2). So it's technically wasteful.

Reply:
I added these constraints so that we can we do not do wasteful computation before cpi into open book. But yeah ideally these kind of errors will be filtered out in testing phase. I can remove these constaints.